File: 5/day5.go
package main

import (
	utils "advent_of_code_2025/pkg/utils"
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	filename := "5/5.txt"
	var fresh [][]int
	var ingredients []int

	inputFile, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error opening input file:", err)
		return
	}
	defer inputFile.Close()

	scanner := bufio.NewScanner(inputFile)

	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "-") {
			from_to := strings.Split(line, "-")
			from, _ := strconv.Atoi(from_to[0])
			to, _ := strconv.Atoi(from_to[1])
			this_range := []int{from, to}
			fresh = append(fresh, this_range)
		} else if line != "" {
			ingredient, _ := strconv.Atoi(line)
			ingredients = append(ingredients, ingredient)
		}
	}

	for {
		last_len := len(fresh)
		for idx1 := range last_len {
			// since fresh can reduce in size
			// skip a loop if indexes are no longer valid
			if idx1 >= len(fresh) {
				continue
			}
			for idx2 := range last_len {
				// skip also when indexes are the same
				if idx2 >= len(fresh) || idx1 == idx2 {
					continue
				}
				// if ranges overlap, drop them and add a new range
				// with bounds that cover the full combined range
				is_overlapping, combined_range := utils.Overlap(fresh[idx1], fresh[idx2])
				if is_overlapping {
					fresh = utils.DropIndexes(fresh, []int{idx1, idx2})
					fresh = append(fresh, combined_range)
				}
			}
		}
		// no change in number of ranges indicates no further
		// range merges are possible
		if len(fresh) == last_len {
			break
		}
	}

	fresh_ids := 0
	for _, rng := range fresh {
		fresh_ids += rng[1] - rng[0] + 1
	}

	n_fresh := 0
	for _, ingredient := range ingredients {
		for _, rng := range fresh {
			if rng[0] <= ingredient && ingredient <= rng[1] {
				n_fresh += 1
				break
			}
		}
	}

	fmt.Print(fresh_ids, "\n")
	fmt.Print(n_fresh, "\n")
}

File: 5/day5.py
def overlap(range1, range2):
    is_overlapping = True
    out_range = []
    if range1[0] > range2[1] or range2[0] > range1[1]:
        is_overlapping = False
    if is_overlapping:
        out_range = [min(range1[0], range2[0]), max(range1[1], range2[1])]
    return is_overlapping, out_range

fresh = []
ingredients = []

with open("5/5.txt") as fs:
    for line in fs.readlines():
        if "-" in line:
            fresh.append([int(n) for n in line[:-1].split("-")])
        elif line != "\n":
            ingredients.append(int(line[:-1]))

while True:
    last_n_fresh = len(fresh)
    for idx in range(last_n_fresh):
        for jdx in range(last_n_fresh):
            if idx != jdx and idx < len(fresh) and jdx < len(fresh):
                is_overlap, new_range = overlap(fresh[idx], fresh[jdx])
                if is_overlap:
                    fresh = [rng for i, rng in enumerate(fresh) if i not in [idx, jdx]]
                    fresh.append(new_range)
    if len(fresh) == last_n_fresh:
        break

print(sum([rng[1] - rng[0] + 1 for rng in fresh]))
print(sum([
    any([rng[0] <= ingredient <= rng[1] for rng in fresh])
    for ingredient in ingredients
]))
